bfgs.go

```diff
@@ -128,34 +121,31 @@ func (b *BFGS) NextDirection(loc *Location, dir []float64) (stepSize float64) {
 		b.first = false
 	}
 
-	// Compute the update rule
-	//     B_{k+1}^-1
-	// First term is just the existing inverse hessian
-	// Second term is
-	//     (sk^T yk + yk^T B_k^-1 yk)(s_k sk_^T) / (sk^T yk)^2
-	// Third term is
-	//     B_k ^-1 y_k sk^T + s_k y_k^T B_k-1
-	//
-	// y_k^T B_k^-1 y_k is a scalar, and the third term is a rank-two update
-	// where B_k^-1 y_k is one vector and s_k is the other. Compute the update
-	// values then actually perform the rank updates.
-	yBy := mat64.Inner(b.yVec, b.invHess, b.yVec)
-	firstTermConst := (sDotY + yBy) / (sDotYSquared)
-	b.tmpVec.MulVec(b.invHess, b.yVec)
-
-	b.invHess.RankTwo(b.invHess, -1/sDotY, b.tmpVec, b.sVec)
-	b.invHess.SymRankOne(b.invHess, firstTermConst, b.sVec)
-
-	// update the bfgs stored data to the new iteration
-	copy(b.x, loc.X)
-	copy(b.grad, loc.Gradient)
-
-	// Compute the new search direction
-	d := mat64.NewVector(b.dim, dir)
-	g := mat64.NewVector(b.dim, loc.Gradient)
-
-	d.MulVec(b.invHess, g) // new direction stored in place
-	floats.Scale(-1, dir)
+	if math.Abs(sDotY) != 0 {
+		// Update the inverse Hessian according to the formula
+		//
+		//  B_{k+1}^-1 = B_k^-1
+		//             + (s_k^T y_k + y_k^T B_k^-1 y_k) / (s_k^T y_k)^2 * (s_k s_k^T)
+		//             - (B_k^-1 y_k s_k^T + s_k y_k^T B_k^-1) / (s_k^T y_k).
+		//
+		// Note that y_k^T B_k^-1 y_k is a scalar, and that the third term is a
+		// rank-two update where B_k^-1 y_k is one vector and s_k is the other.
+		yBy := mat64.Inner(&b.y, b.invHess, &b.y)
+		b.tmp.MulVec(b.invHess, &b.y)
+		scale := (sDotY + yBy) / sDotY / sDotY
```

Comment:
Excellent and depressing suggestion at the same time: it makes that test pass, otherwise the tolerance would have to be changed from 1e-10 to 1e-8!